### ISSUE TITLE
[Help wanted][Rebalance events] Sporadic errors when performing actor calls

### ISSUE THREAD

**Issue by KrylixZA** [dapr/dapr#7110]
## In what area(s)?

/area runtime
/area placement

## What version of Dapr?

> 1.11.x (any version of 1.11 exhibited these errors too)
> 1.12.0

## Expected Behavior

No errors to occur unless there is a genuine outage or environmental error in the system.

## Actual Behavior

The system we have built processes fairly large volumes of data through our actors (average of 5m events per hour), hosted in Azure Kubernetes Service with Azure Cosmos DB. The behaviour we are seeing are regular blips in errors **_when calling the actors through the actor proxy_**. See below for a graph of our App Insights errors:

![image](https://github.com/dapr/dapr/assets/16819506/4e7e9b49-07f0-4136-8186-ad22eab6073d)

We've investigated these issues and they appear to be related to scale events on our pods and often times the VMs that sit underneath the pods as well. Thankfully, the service is triggered by Pub/Sub which allows us to leverage the retry and re-process behaviour built into Dapr to guarantee message processing. That said, we still want to iron out any "known errors" - this is one of them that we can't eliminate. 
This issue in particular seems to be related to the placement service not being aware of scale events - which makes sense. In Dapr 1.11, this issue would be quick to resolve (a couple milliseconds at most). However, since upgrading to Dapr 1.12, we now seem to be waiting 20s for the same error to resolve. The 20s timeout seems like it could be related to the following change released in Dapr 1.12:

![image](https://github.com/dapr/dapr/assets/16819506/1799c010-1270-471e-b54f-37d5d4ce54d6)

We upgraded to 1.12 as we felt it was more stable than 1.11 and promised benefits. However, we have a strict SLA on our event processing which means this now less common but more expensive error is becoming a real problem for us. We've theocratised about introducing changes to our resiliency policy to cut off calls to the actor after 1s and discard the message off the topic (and potentially push to a dead letter topic) but are reluctant to make this change as it's a great advertising point to guarantee message processing and something we pride ourselves on.

I am logging this issue with the hopes of this issue being identified and resolved in the longer term, but also seeking advice on how to better handle this situation as a temporary workaround. 

TIA 💯 

## Steps to Reproduce the Problem

It's very difficult as it requires scale events on the pods and in particular requires relatively high throughput to trigger the errors, as the events need to come through faster than the placement service updates its placement tables and inform actor services of the change. Outside of production loads, we have not been able to reproduce this issue.
Currently we are using HPA in production but will be switching to KEDA in the near future. Unfortunately there is no way for KEDA to inform the placement service that actors need to be rebalanced. 

## Release Note

RELEASE NOTE:

**Comment by KrylixZA** [comment]
As a sidebar thing, it would be cool if we could configure a KEDA scaled object to inform the placement service of scale events. I haven't quite thought through how that might work but conceptually it sounds like a cool idea.

**Comment by ItalyPaleAle** [comment]
Could you share more details on the errors you are seeing? What do the Dapr logs say (and possibly the placement logs too)?

As for changes in 1.12, there have been quite a few, all aimed at making actors more stable as you mentioned.

I am not sure that specific change from the changelog you highlight would be causing this issue, however. If anything, that change was meant to _prevent_ errors from appearing when your app is invoking a Dapr API that interacts with actors right after Dapr has started up.

What this may be more related to _could_ be #6966, which was an important bug fix. Previously, as soon as an actor host, Dapr considered it healthy and ready to host actors, before invoking `/healthz` on the app. Now, it requires the app to perform at least 1 successful healthcheck before it considers it as healthy.

**Comment by KrylixZA** [comment]
Hey @ItalyPaleAle.
 
Below is a sanitised .NET stack trace from the running application. Notably, there are no logs from the sidecar at that particular timestamp. The sidecar is using the default log level of `info`.

**Stack trace:**
> 2023-10-31 05:30:38.3125|ERROR|GenericMiddleware.ErrorHandler|error invoke actor method: rpc error: code = Canceled desc = grpc: the client connection is closing|Exception occurred:Dapr.DaprApiException: error invoke actor method: rpc error: code = Canceled desc = grpc: the client connection is closing
   at Dapr.Actors.DaprHttpInteractor.SendAsyncHandleUnsuccessfulResponse(Func`1 requestFunc, String relativeUri, CancellationToken cancellationToken)
   at Dapr.Actors.DaprHttpInteractor.SendAsync(Func`1 requestFunc, String relativeUri, CancellationToken cancellationToken)
   at Dapr.Actors.DaprHttpInteractor.InvokeActorMethodWithRemotingAsync(ActorMessageSerializersManager serializersManager, IActorRequestMessage remotingRequestRequestMessage, CancellationToken cancellationToken)
   at Dapr.Actors.Communication.Client.ActorRemotingClient.InvokeAsync(IActorRequestMessage remotingRequestMessage, CancellationToken cancellationToken)
   at Dapr.Actors.Client.ActorProxy.InvokeMethodAsync(Int32 interfaceId, Int32 methodId, String methodName, IActorRequestMessageBody requestMsgBodyValue, CancellationToken cancellationToken)
   at ActorManager.SendEventToActorAsync(String modelId, Event playerEvent, EventType eventType, BufferConfiguration bufferSizes, IActor playerEventsActor, Guid correlationId, CancellationToken cancellationToken) in /app/src/ActorManager.cs:line 206
   at ActorManager.ProcessEventAsync(String modelId, Event playerEvent, EventType eventType, Guid correlationId, CancellationToken cancellationToken) in /app/src/ActorManager.cs:line 135
   at Controller.ProcessEventAsync(PubSubDocument`1 wagerEvent, CancellationToken cancellationToken) in /app/src/Controller.cs:line 158
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at TelemetryHttpContext.InvokeAsync(HttpContext context) in /app/src/TelemetryHttpContext.cs:line 63
   at GenericMiddleware.ErrorHandler.InvokeAsync(HttpContext context)

You mentioned this which immediately triggered a lightbulb moment for me. In our Helm chart, the readiness probe is configured as follows:
```yaml
{{- if .Values.healthProbes.readinessEnabled }}
readinessProbe:
  httpGet:
    path: /healthz
    port: {{ .Values.http.listenPort }}
  initialDelaySeconds: 20
  periodSeconds: 10
{{- end }}
```

That `initialDelaySeconds: 20` is exactly the timeout what we're seeing. I'll experiment with dropping that number to at least force the call to fail quicker. Obviously that's not a perfect solution but might serve as a decent workaround for the time being.

**Comment by ItalyPaleAle** [comment]
The readiness probe should not really impact Dapr, as we have our own healthchecks. If that's the case, it's very interesting.

From the logs, the most interesting thing is that the gRPC connection is dropped.

I assume for you running Dapr with debug logs on would not be an option, would it?

Also, do you know if there's any event in K8s at the time? Are the pods perhaps crashing?

**Comment by KrylixZA** [comment]
> I assume for you running Dapr with debug logs on would not be an option, would it?

It can be an option. Will try get a test run in the next day or two 👍 

> Also, do you know if there's any event in K8s at the time? Are the pods perhaps crashing?

Typically our pods are very stable. We don't often have pods crash, particular the pods that interact with actors. That said, there are k8s scale events happening all the time as our volumes have moderate to high volatility. These events are combinations of HPA scale events on the pods and node scale events on the underlying VM Scale Set. The node scale events are easy to identify because we set a hard limit of 100 to the number of running pods per VM and we see errors in multiples of 100 when they scale. The ones that are difficult to trace are the pod scale events that are not any particular amount. We are confident it is related to scale events, but have no definitive way of proving that using HPA (as it the system is configured at the moment). With KEDA we're going to start pulling pod count metrics through Prometheus which would give us accurate correlations between pod scale events and these errors.

**Comment by KrylixZA** [comment]
Just a quick update, over the last 24 hours we've only seen error counts in 100s or multiples thereof. These we know are related to node scale events on the underlying VM Scale Set. I am going to monitor this behaviour of the next 24 hours to see if this remains consistent. If so, these errors are purely related to node shutdowns in the VM Scale Set and we'll look to optimise our IaC. 

We have taken on more traffic into our system in the last week. Since then it appears the errors are largely multiples of 100. My working theory now is that these errors have been related to node scale events all along but we've just not had enough volume for it to be easily identifiable.

Will feedback to this issue tomorrow 👍

**Comment by siyakhumalo-ms** [comment]
> Just a quick update, over the last 24 hours we've only seen error counts in 100s or multiples thereof. These we know are related to node scale events on the underlying VM Scale Set. I am going to monitor this behaviour of the next 24 hours to see if this remains consistent. If so, these errors are purely related to node shutdowns in the VM Scale Set and we'll look to optimise our IaC.
> 
> We have taken on more traffic into our system in the last week. Since then it appears the errors are largely multiples of 100. My working theory now is that these errors have been related to node scale events all along but we've just not had enough volume for it to be easily identifiable.
> 
> Will feedback to this issue tomorrow 👍

Hey @KrylixZA, just following up on this issue.

**Comment by KrylixZA** [comment]
Hey all.

Sorry for the late feedback. I can confirm these errors are all related to node shutdowns resulting from scale-down events on the pods. 

In our case, this particular error is always transient. The service requeues the message to the Kafka topic after receiving the 500 response code from the actor proxy, after which the message is re-processed successfully 100% of the time. This would suggest, then, that some kind of retry mechanism in the actor proxy to point to the new address of the actor after the initial failure would be a workaround to this problem. With the current resiliency policy, I am not sure this is possible. As best I understand, the actor retry policy is not a retry on the call to the actor, but rather the actor itself being able to retry. It would be nice if there were a way for the actor proxy to retry a certain amount of times.

Alternatively, a way to inform the placement service of scale events (for either node or pods) in the cluster could potentially be a better solution to this problem as ultimately it comes down to the fact that the placement service has an address for an actor which is no longer valid.

Are any of the above two options viable at present? 

TIA 💯

**Comment by KrylixZA** [comment]
To all reading this, we've made some progress on this our side. It might seem quite obvious but also is likely to trip others up in the future.

The errors we saw on the actor calls are largely eliminated now. We achieved this simply by boosting the minimum number of pods available at any given time. When we initially went live, we were using HPA with a minimum of 2 pods and maximum of 100 pods to allow the system to scale into the loads. HPA wasn't perfect but it did the job, albeit with the errors as documented in the OP.

We have recently switched to KEDA and set the consumer lag threshold for Kafka to a relatively low number given the message throughput which essentially meant the deployment was elastically scaling very aggressively which actually caused this error to happen a lot more often in a given time frame than before, however the overall throughput was higher and consumer lag remained lower, relative to HPA. So we lived with it.

However, after watching how KEDA behaves for a while, we boosted the minimum number of pods and threshold values to what their respective averages and now the errors have gone away completely. 

I can, therefore, confirm this error is related to node scale down events in AKS which result in the placement service incorrectly informing the actor proxy of where the actor actually lives. 

The last remaining "known dapr error" we are still seeing is the same as the one documented on this issue: https://github.com/microsoft/azure-container-apps/issues/294.

I am unsure whether to close this issue or not as I do strongly believe that the Dapr placement service should expose or subscribe to node scale events on Kubernetes clusters as this inherently means actors will rebalance and move around. The errors are misleading and hide real issues in clusters. Some teams may also have a genuine need for scale-to-zero approach to their actor host services as well which would suffer the same fate.

**Comment by KrylixZA** [comment]
Howdy folks. Looping back here. We are still seeing these errors calling the actors. The only way to practically avoid this entirely is to have an over-provisioned Kubernetes cluster that never scales down. 

When I set the minimum number of pods to 30 a couple weeks ago, this was enough to handle the volumes. However, we are entering a peak period as the year draws to a close and have started to see the actor failures once again as the cluster scales out to match the volumes, and subsequently scales down during quieter periods. There appears to be no way around this error while leveraging auto-scaling as a concept. We are also still seeing the 404 errors on the `GET /dapr/config` error that is raised against Azure Container Apps.

These errors are becoming "known errors" which is a concept that makes my skin crawl, but is what it is right now. 

PS: The reason we were seeing errors in multiples of 100s had nothing to do with pods per node constraints. It comes down to strange behaviour in Application Insights error counts when you have very low sampling percentages (currently we are only sampling 1% of all events due to volumes and costs).

**Comment by yaron2** [comment]
This is being investigated and worked on.

**Comment by KrylixZA** [comment]
Seems to be related to https://github.com/microsoft/azure-container-apps/issues/1070

**Comment by KrylixZA** [comment]
Looks like this issue has been resolved in [Dapr 1.13.2](https://github.com/dapr/dapr/releases/tag/v1.13.2). Ultimately we were publishing events to the next Kafka topic for processing out of the actor as the final step in it's processing flow. Between actor rebalancing, cluster rebalancing, auto-scaling and new deployments, cancellations were being requested while the actor was still trying to publish the event. With the fixes released in 1.13.2, this is now resolved.
